wasi-common/src/pipe.rs

```diff
@@ -109,8 +111,21 @@ impl<R: Read + Any + Send + Sync> WasiFile for ReadPipe<R> {
         Ok(FileType::Pipe)
     }
     async fn read_vectored<'a>(&mut self, bufs: &mut [io::IoSliceMut<'a>]) -> Result<u64, Error> {
-        let n = self.borrow().read_vectored(bufs)?;
-        Ok(n.try_into()?)
+        let n = self.borrow().read_vectored(bufs)?.try_into()?;
+        self.offset += n;
+        Ok(n)
+    }
+    async fn read_vectored_at<'a>(
+        &mut self,
+        bufs: &mut [io::IoSliceMut<'a>],
+        offset: u64,
+    ) -> Result<u64, Error> {
+        if offset != self.offset {
+            return Err(Error::badf());
+        }
```

Comment:
Is there a use case that needs `read_vectored_at` on `ReadPipe`? In https://github.com/bytecodealliance/preview2-prototyping/pull/29, I'm working to split a `WasiStream` out of `WasiFile`, and I haven't been anticipating `WasiStream` would need the `_at` forms of read/write, because streams aren't random-access.

Reply:
It's because there's currently no `read` function in wasi-filesystem -- only `pread`.  `fd_read` in the adapter uses `pread`, which I assumed was intentional.

Reply:
It is intentional, but it's not very developed yet. Would you mind reverting this change to add an offset here? It's my goal that with #29 it should be clearer how to do this.